hummingbot/connector/derivative/phemex_perpetual/phemex_perpetual_api_order_book_data_source.py

```diff
@@ -0,0 +1,233 @@
+            rate=Decimal(symbol_info["fundingRateRr"]),
+        )
+        return funding_info
+
+    async def _request_order_book_snapshot(self, trading_pair: str) -> Dict[str, Any]:
+        ex_trading_pair = await self._connector.exchange_symbol_associated_to_pair(trading_pair=trading_pair)
+
+        params = {"symbol": ex_trading_pair}
+
+        data = await self._connector._api_get(path_url=CONSTANTS.SNAPSHOT_REST_URL, params=params)
+        return data
+
+    async def _order_book_snapshot(self, trading_pair: str) -> OrderBookMessage:
+        snapshot_response: Dict[str, Any] = (await self._request_order_book_snapshot(trading_pair))["result"]
+        snapshot_msg: OrderBookMessage = OrderBookMessage(
+            OrderBookMessageType.SNAPSHOT,
+            {
+                "trading_pair": trading_pair,
+                "update_id": snapshot_response["sequence"],
+                "bids": snapshot_response[self._diff_messages_queue_key]["bids"],
+                "asks": snapshot_response[self._diff_messages_queue_key]["asks"],
+            },
+            timestamp=snapshot_response["timestamp"] // 1e9,
+        )
+        return snapshot_msg
+
+    async def _connected_websocket_assistant(self) -> WSAssistant:
+        url = web_utils.wss_url(CONSTANTS.PUBLIC_WS_ENDPOINT, self._domain)
+        ws: WSAssistant = await self._api_factory.get_ws_assistant()
+        async with self._api_factory.throttler.execute_task(limit_id=CONSTANTS.WSS_CONNECTION_LIMIT_ID):
+            await ws.connect(ws_url=url, ping_timeout=CONSTANTS.WS_HEARTBEAT)
+        return ws
+
+    async def _subscribe_channels(self, ws: WSAssistant):
+        """
+        Subscribes to the trade events and diff orders events through the provided websocket connection.
+        :param ws: the websocket assistant used to connect to the exchange
+        """
+        try:
+            stream_id_channel_pairs = [
+                CONSTANTS.DIFF_STREAM_METHOD,
+                CONSTANTS.TRADE_STREAM_METHOD,
+                CONSTANTS.FUNDING_INFO_STREAM_METHOD,
+            ]
+            for stream_method in stream_id_channel_pairs:
+                params = []
+                for trading_pair in self._trading_pairs:
+                    params.append(await self._connector.exchange_symbol_associated_to_pair(trading_pair=trading_pair))
+                payload = {
+                    "id": 0,
+                    "method": stream_method,
+                    "params": params if stream_method is not CONSTANTS.FUNDING_INFO_STREAM_METHOD else [],
+                }
+                subscribe_request: WSJSONRequest = WSJSONRequest(payload)
+                async with self._api_factory.throttler.execute_task(limit_id=CONSTANTS.WSS_MESSAGE_LIMIT_ID):
+                    await ws.send(subscribe_request)
+            self.logger().info("Subscribed to public order book, trade and funding info channels...")
+            safe_ensure_future(self.ping_loop(ws=ws))
```

Comment:
We don't subscribe to ticker info to provide "shallow order book" functionality. In order to do that, we also need to remove the diffs subscription if shallow order book is activated. Bybit has an order book data source that supports shallow order books.